tests/test_gf_unpack2.F90

```diff
@@ -4,197 +4,192 @@
+       & char(0), char(0), char(0), char(0), char(0),  char(0),   &
+                                ! section 8
+       & "7", "7", "7", "7" /) 
+
+  print *, ''//NEW_LINE('A')//'Calling grib_info ...'
+  call gribinfo(fgrib, fgrib_len, listsec0, listsec1,  &
+       numlocal, numfields, maxvals, ierr)
+  print *, 'Message length: ', listsec0(3)
+  print *, 'Section 2 occurence: ', numlocal
+  print *, 'Section 2 length: ', maxvals(1)
+
+
+  print *, ''//NEW_LINE('A')//'Testging gf_unpack2 ...'
+
+  ! Normal test, should return no error and allocate csec2
+  ! Start byte = 37 => start bit = 296
+  ! Length of section (in octets) is 6
+  iofst = 296
+  call gf_unpack2(fgrib, fgrib_len, iofst, lencsec2, csec2, ierr)
+  if (ierr .ne. 0 .or. iofst .ne. 384 .or. lencsec2 .ne. 6) stop 1
+  deallocate(csec2)
+
+  ! Offset not to section 2 data, should error without allocating csec2
+  iofst = 1
+  call gf_unpack2(fgrib, fgrib_len, iofst, lencsec2, csec2, ierr)
+  if (ierr .ne. 6) stop 2
+
+  ! Offset to section 2 data with already allocated csec2 array
+  ! should error and nullify csec2
+  iofst = 296
+  allocate(csec2(100))
+  call gf_unpack2(fgrib, fgrib_len, iofst, lencsec2, csec2, ierr)
+  print *,'ierr: ', ierr
+  if (ierr .ne. 6) stop 3
+
+  ! ! Printing all offsets which return as section 2 in gf_unpack2
+  ! print *,''//NEW_LINE('A')//'Offsets that give section 2:'
+
+  ! do i=1, 465
+  !    iofst = i
+  !    ierr=0
+  !    lencsec2=0
+  !    nullify(csec2)
+
+  !    call g2_gbytec(fgrib,lensec,iofst,32)        ! Get Length of Section
+  !    iofst=iofst+32    
+  !    lencsec2=lensec-5
+  !    call g2_gbytec(fgrib,isecnum,iofst,8)         ! Get Section Number
+  !    iofst=iofst+8   
+  !    iofst=iofst+(lencsec2*8)
+
+  !    if (isecnum .eq. 2) print *, 'Offset: ', i, 'length: ', lencsec2
+
+  ! end do
+
+  ! print *, ''//NEW_LINE('A')//'Confirming lengths with gf_unpack2:'
+  ! iofst = 81
+  ! i = iofst
+  ! call gf_unpack2(fgrib, fgrib_len, iofst, lencsec2, csec2, ierr)
+  ! deallocate(csec2)
+  ! print *, 'Offset: ', i, 'length: ', lencsec2
+  ! iofst = 129
+  ! i = iofst
+  ! call gf_unpack2(fgrib, fgrib_len, iofst, lencsec2, csec2, ierr)
+  ! deallocate(csec2)
+  ! print *, 'Offset: ', i, 'length: ', lencsec2
+  ! iofst = 185
+  ! i = iofst
+  ! call gf_unpack2(fgrib, fgrib_len, iofst, lencsec2, csec2, ierr)
+  ! deallocate(csec2)
+  ! print *, 'Offset: ', i, 'length: ', lencsec2
+  ! iofst = 296
+  ! i = iofst
+  ! call gf_unpack2(fgrib, fgrib_len, iofst, lencsec2, csec2, ierr)
+  ! deallocate(csec2)
+  ! print *, 'Offset: ', i, 'length: ', lencsec2
+  ! iofst = 465
+  ! i = iofst
+  ! call gf_unpack2(fgrib, fgrib_len, iofst, lencsec2, csec2, ierr)
+  ! deallocate(csec2)
+  ! print *, 'Offset: ', i, 'length: ', lencsec2
+
+  ! print *, ''//NEW_LINE('A')//'Calling gf_unpack with offset 24 ...'
+  ! iofst = 24
+  ! call gf_unpack2(fgrib, fgrib_len, iofst, lencsec2, csec2, ierr)
```

Comment:
I still have a question here, is the user expected to know the correct offset and it's unlikely that the wrong offset would be used? If you choose the offset carefully you can get the program to think it's unpacking section 2 but the length is bogus and causes a seg fault. Not sure if it's significant but I thought I would bring it up.

Reply:
Correct, the caller just has to know the correct offset for section 2. Calling with an incorrect offset which coincidentally falls on a patch of data which looks like the start of section 2 will do nutty things.